### PR TITLE
Moving average with offset clause

### LINKED ISSUE #46669: Add moving average option to "compare to the past" modal

### PR DESCRIPTION
**Pull request by romeovs**
Closes https://github.com/metabase/metabase/issues/46669

### Description

Adds the option to calculate the x-period moving average or difference (absolute or %) vs the moving average in the `Compare to past` menu.

### How to verify

Describe the steps to verify that the changes are working as expected.

1. New Question -> Orders
2. Aggregation -> Count
3. Aggregation -> Compare to past
4. Select `Moving average`
5. Click `Done`
6. Moving average columns should be added to the query

### Demo

https://github.com/user-attachments/assets/860dc09e-3437-45dc-8c5c-cdf618f2b70a

### Checklist

- Unit test have been added, I will add e2e test for this feature in a follow up PR for https://github.com/metabase/metabase/issues/46531, since they will change considerably.

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/CompareAggregations/components/ColumnPicker/ColumnPicker.tsx [hunk 1]
```diff
@@ -94,3 +82,62 @@ const Item = forwardRef<
...
+        value: "diff-moving-average",
+      },
+    ];
+  }
+  return [];
+}
+
+function convertValues(
+  values: ColumnType[],
+  comparisonType: string,
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/CompareAggregations/components/ColumnPicker/ColumnPicker.tsx:130]: enum?

File: frontend/src/metabase/query_builder/components/CompareAggregations/components/ColumnPicker/ColumnPicker.tsx [hunk 2]
```diff
@@ -94,3 +82,62 @@ const Item = forwardRef<
...
+): ColumnType[] {
+  if (comparisonType === "offset") {
+    return values.map(
+      value => value.replace("moving-average", "offset") as ColumnType,
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/CompareAggregations/components/ColumnPicker/ColumnPicker.tsx:134]: Issues here won't be caught with typescript. Can we switch to a `Record<ColumnType, ColumnType>` or something like this to make sure it's type-safe?

File: frontend/src/metabase/query_builder/components/CompareAggregations/components/ColumnPicker/ColumnPicker.tsx [hunk 3]
```diff
@@ -15,39 +16,26 @@ interface ItemType {
...
-export const ColumnPicker = ({ value, onChange }: Props) => {
+export const ColumnPicker = ({ value, onChange, comparisonType }: Props) => {
   const handleChange = useCallback(
     (values: string[]) => {
       onChange(values as ColumnType[]);
     },
     [onChange],
   );
 
+  useDeepCompareEffect(() => {
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/CompareAggregations/components/ColumnPicker/ColumnPicker.tsx:32]: effects are usually hard to follow. Can we do this conversion of column types in event handlers instead?

File: frontend/src/metabase-lib/moving-average.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+  return withExpressionName(newClause, newName);
+}
+
+function getMovingAverageClauseName(
+  query: Query,
+  stageIndex: number,
+  clause: AggregationClause | ExpressionClause,
+  offset: number,
+  _includeCurrentPeriod: boolean,
```
Comment (ranquild) [inline frontend/src/metabase-lib/moving-average.ts:107]: It's a private function. Maybe remove this if not needed?

File: frontend/src/metabase-lib/moving-average.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+  const newClause = expressionClause("/", [
+    expressionClause("+", clauses),
+    Math.abs(offset),
+  ]);
+
+  return newClause;
+}
+
+export function movingAverageClause(
```
Comment (ranquild) [inline frontend/src/metabase-lib/moving-average.ts:34]: maybe use an object instead, like we do for `...filterParts` etc? `MovingAverageClauseParts`

File: frontend/src/metabase-lib/moving-average.unit.spec.ts [hunk 6]
```diff
@@ -0,0 +1,1132 @@
...
+          offset,
+          includeCurrentPeriod,
+        );
+        it("produces correct aggregation name", () => {
+          const info = displayInfo(query, stageIndex, clause);
+          expect(info.displayName).toBe("Count (2-period moving average)");
+        });
+
+        it("produces correct aggregation clause", () => {
+          expect(toLegacyQuery(query)).toMatchObject({
```
Comment (ranquild) [inline frontend/src/metabase-lib/moving-average.unit.spec.ts:91]: Any direct MBQL should be avoided. Please use `expressionParts` with `expect(...).toMatchObject`

File: frontend/src/metabase/query_builder/components/CompareAggregations/CompareAggregations.tsx [hunk 7]
```diff
@@ -41,13 +44,27 @@ export const CompareAggregations = ({
...
 
   const title = useMemo(
     () => getTitle(query, stageIndex, aggregation),
     [query, stageIndex, aggregation],
   );
 
+  const handleComparisonTypeChange = useCallback(
+    (comparisonType: ComparisonType) => {
+      setComparisonType(comparisonType);
+      if (comparisonType === "moving-average" && offset === 1) {
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/CompareAggregations/CompareAggregations.tsx:61]: Screenshot 2024-08-13 at 10 28 36

Try changing it to `1` afterwards. I think we should coerce it to 2 automatically and don't show this error.

**Review by ranquild — APPROVED** [review]
Great work! Some small cleanup needed

File: frontend/src/metabase-lib/moving-average.ts (hunk 5 above)
Comment (romeovs) [reply frontend/src/metabase-lib/moving-average.ts:34]: I did it like this to be consistent [with Kamil's work on offsets](https://github.com/metabase/metabase/blob/master/frontend/src/metabase-lib/offset.ts#L10). What's the benefit of using `...Parts`, if we want to turn it into a clause afterwards?

Adding a display name to expression parts is not possible it seems and building and expression clause from `ExpressionParts` using `Lib.expressionClause` is also showing TypeScript errors for me.

Comment (ranquild) [reply frontend/src/metabase-lib/moving-average.ts:34]: Parts was just a convention for an interface name. So instead of passing 5 args you would pass an object. There is nothing else being suggested here.

Comment (romeovs) [reply frontend/src/metabase-lib/moving-average.ts]: Ah, yeah that makes sense!
